monai/utils/component_store.py

```diff
@@ -0,0 +1,117 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from collections import namedtuple
+from keyword import iskeyword
+from textwrap import dedent, indent
+from typing import Any, Callable, Iterable, TypeVar
+
+T = TypeVar("T")
+
+
+def is_variable(name):
+    """Returns True if `name` is a valid Python variable name and also not a keyword."""
+    return name.isidentifier() and not iskeyword(name)
+
+
+class ComponentStore:
```

Comment:
we might be able to use/enhance the existing `LayerFactory` to achieve the same capability https://github.com/Project-MONAI/MONAI/blob/0a3cdd9f1403cbad057b96230840ee17cf430dbf/monai/networks/layers/factories.py#L76 

cc @ericspod I think the usage is https://github.com/Project-MONAI/GenerativeModels/blob/f28afa003f9b5be0ebf9f31270abc06e7957a607/generative/networks/schedulers/scheduler.py#L44

Reply:
> we might be able to use/enhance the existing `LayerFactory` to achieve the same capability

I thought about this before when writing this initially but concluded that there were too many differences between the two classes. We could refactor things such that they inherit from a common implementation, we need two classes because the usage here isn't to generate layers but be a factory for functions.

Reply:
But again these aren't network layers. The purpose of LayerFactory is to be a factory for the components of a network, so at the least we need to rename classes a bit.

Reply:
indeed, the `LayerFactory` doesn't assume network components/blocks, the class could be renamed to a more generic one and deprecating the existing one. or have a refactoring into two subclasses as @ericspod mentioned. would you be interested in implementing it? @marksgraham

Reply:
so a `Factory` class which `LayerFactory` and `ComponentStore` inherit from? That to me seems preferable to deprecating `LayerFactory`.

Reply:
sounds good, thanks @marksgraham!